sw/host/opentitanlib/src/console/spi.rs

```diff
@@ -0,0 +1,51 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::Result;
+use std::cell::Cell;
+use std::time::Duration;
+
+use crate::io::console::ConsoleDevice;
+use crate::io::spi::{Target, Transfer};
+
+pub struct SpiConsoleDevice<'a> {
+    pub spi: &'a dyn Target,
+    pub console_next_frame_number: Cell<u32>,
+}
+
+impl<'a> SpiConsoleDevice<'a> {
+    const SPI_FRAME_HEADER_SIZE: usize = 8;
+    const SPI_MAX_DATA_LENGTH: usize = 2032;
+}
+
+impl<'a> ConsoleDevice for SpiConsoleDevice<'a> {
+    fn console_read(&self, buf: &mut [u8], _timeout: Duration) -> Result<usize> {
+        // Read the SPI console frame header.
+        let mut header = vec![0u8; SpiConsoleDevice::SPI_FRAME_HEADER_SIZE];
+        self.spi
+            .run_transaction(&mut [Transfer::Write(&[0xff; 4]), Transfer::Read(&mut header)])?;
+        let frame_number: u32 = u32::from_le_bytes(*&header[0..4].try_into().unwrap());
+        let data_len_bytes: usize = u32::from_le_bytes(*&header[4..8].try_into().unwrap()) as usize;
+        if frame_number != self.console_next_frame_number.get()
+            || data_len_bytes > SpiConsoleDevice::SPI_MAX_DATA_LENGTH
+        {
+            // This frame is junk, so we do not read the data.
+            return Ok(0);
+        }
+        self.console_next_frame_number.set(frame_number + 1);
+
+        // Read the SPI console frame data.
+        let data_len_bytes_w_pad = data_len_bytes + (4 - (data_len_bytes % 4));
+        let mut data = vec![0u8; data_len_bytes_w_pad];
+        self.spi
+            .run_transaction(&mut [Transfer::Write(&[0xff; 4]), Transfer::Read(&mut data)])?;
```

Comment:
Does the firmware implementation always write the header and the data frame in "atomically" with respect to the upstream SPI host's view?

Reply:
yes, both header and data are written in together at the same time, "atomically", see #17818